src/RevProc.cc

```diff
@@ -1581,45 +1801,89 @@ bool RevProc::ClockTick( SST::Cycle_t currentCycle ){
   // else if the the instruction has not yet been triggered, execute it
   // else, wait until the counter is decremented to zero to retire the instruction
   //
-  for (int tID = 0; tID < _REV_THREAD_COUNT_; tID++){
-    THREAD_CTS[tID] = (RegFile[tID].cost == 0);
+  //
+  if( PendingCtxSwitch ){
+    /*
+     * There was a ctx switch event triggered
+     * - Either a call to fork/clone
+     * - Child process finished executing 
+     */
+    if( Pipeline.empty() ) {
+      if( !ChangeActivePID(NextPID) ){
+        output->fatal(CALL_INFO, 0, 0, 
+                      "Core %d ; Hart %u; PID %d Failed to change active PID to %u\n",
+                      id, HartToDecode, GetActivePID(), NextPID);
+      } else {
+        RegFile->trigger = 0;
+        RegFile->cost = 0;
+        ExecPC = GetPC();
+        PendingCtxSwitch = false;
+        NextPID = 0;
+      }
+    }
   }
 
-  if( THREAD_CTS.any() && (!Halted)) { // && (RegFile[threadID].cost == 0)){
+  for (int tID = 0; tID < _REV_HART_COUNT_; tID++){
+    HART_CTS[tID] = (GetRegFile(tID)->cost == 0);
+  }
+
+  if( HART_CTS.any() && (!Halted)) { // && (RegFile(HartID)->cost == 0)){
     // fetch the next instruction
     ResetInst(&Inst);
 
     //Determine the active thread
-    threadToDecode = GetThreadID();
+    HartToDecode = GetHartID();
+
+    if( !PrefetchInst() ){
+      Stalled = true;
+      Stats.cyclesStalled++;
+    }else{
+      Stalled = false;
+    }
 
     // If the next instruction is our special bounce address
     // DO NOT decode it.  It will decode to a bogus instruction.
     // We do not want to retire this instruction until we're ready
-    if( GetPC() != _PAN_FWARE_JUMP_ ){
+    if( (GetPC() != _PAN_FWARE_JUMP_) && (!Stalled) ){
       Inst = DecodeInst();
+      Inst.entry = RegFile->Entry;
     }
+
+    //Now that we have decoded the instruction, check for pipeline hazards
+    if(Stalled || DependencyCheck(HartToDecode, &Inst)) {
+      RegFile->cost = 0; // We failed dependency check, so set cost to 0 - this will
+      Stats.cyclesIdle_Pipeline++;        // prevent the instruction from advancing to the next stage
+      HART_CTE[HartToDecode] = false;
+      HartToExec = _REV_INVALID_HART_ID_;
+    }else {                 
+      Stats.cyclesBusy++;
+      HART_CTE[HartToDecode] = true;
+      HartToExec = HartToDecode;
+    };
+    Inst.cost = RegFile->cost;
```

Comment:
I believe this should be InstTable[Inst.entry].cost

Reply:
RHS of the operation?  The cost is set during decode in order to track the lifetime of the pipeline cost. Maybe I'm missing something though.

Reply:
Initially I was worried that the RegFile.cost may not be updated correctly, but I ran a few tests and it appears to work correctly, so I'm good with leaving as-is.